### PR TITLE
logging and better result propagation

### PR DESCRIPTION
**Pull request by agend**

### CODE REVIEW

File: src/bin/bobd.rs [hunk 1]
```diff
@@ -32,26 +38,52 @@ impl server::BobApi for BobSrv {
...
-                    let mut opts: BobOptions = Default::default();
-                    if let Some(vopts) = param.options.as_ref() { 
-                        if vopts.force_node {
-                            opts |= BobOptions::FORCE_NODE;
+            let sw = Stopwatch::start_new();
+            let param = req.into_inner();
+
+            if !self.put_validate(&param) {
+                warn!("PUT[-] invalid arguments - key and data is mandatory");
+                Box::new(future::err(tower_grpc::Status::new(tower_grpc::Code::InvalidArgument, "Key and data is mandatory")))
```
Comment (ikopylov) [inline src/bin/bobd.rs]: I would prefer to see the `return` expression here. It is easy to make a mistake with this functional style by  simply adding a code after the `if` statement.

File: src/bin/bobd.rs [hunk 2]
```diff
@@ -32,26 +38,52 @@ impl server::BobApi for BobSrv {
...
-                    opts
-                }
-            ).then(|_r| ok(Response::new(OpStatus {
-                error: None
-            }))))
+                ).then(move |r| {
+                    let elapsed = sw.elapsed_ms();
+                    match r {
+                        Ok(r_ok) => {
+                            info!("PUT[{}]-OK local:{} ok dt: {}ms", key, !r_ok.is_clustered, elapsed);
```
Comment (ikopylov) [inline src/bin/bobd.rs]: This message will be written for every successful request. It is better to write it at the 'debug' level.

File: src/bin/bobd.rs [hunk 3]
```diff
@@ -32,26 +38,52 @@ impl server::BobApi for BobSrv {
...
+                Box::new(future::err(tower_grpc::Status::new(tower_grpc::Code::InvalidArgument, "Key and data is mandatory")))
+            } else {
+                let key = BobKey{
+                            key: param.key.unwrap().key
+                        };
+                let data = BobData{
+                            data: param.data.unwrap().data
+                        };
+
+                debug!("PUT[{}] data size: {}", key, data.data.len());
```
Comment (ikopylov) [inline src/bin/bobd.rs]: It is better to use 'trace' level here

File: src/core/bob_client.rs [hunk 4]
```diff
@@ -67,11 +67,14 @@ impl BobClient {
...
-            println!("ERR = {:?}", e);
+            debug!("BobClient: ERR = {:?}", e);
         })
     }
 
-    pub fn put(&mut self, key: BobKey, data: &BobData) -> impl Future<Item=BobPutResult, Error=BobError> {
+    pub fn put(&mut self, key: BobKey, data: &BobData)
+            -> impl Future<Item=ClusterResult<BobPutResult>, Error=ClusterResult<BobError>> {
+        let n1 = self.node.clone();
+        let n2 = self.node.clone();
```
Comment (ikopylov) [inline src/core/bob_client.rs]: Why do you clone it twice in the beginning of method?

File: src/bin/bobd.rs (hunk 1 above)
Comment (agend) [reply src/bin/bobd.rs]: This is rust style) and clippy (rust linter) will generate warning of needless return

File: src/bin/bobd.rs (hunk 2 above)
Comment (agend) [reply src/bin/bobd.rs]: this is intentional now

File: src/bin/bobd.rs (hunk 3 above)
Comment (agend) [reply src/bin/bobd.rs]: make sense

File: src/core/bob_client.rs (hunk 4 above)
Comment (agend) [reply src/core/bob_client.rs]: as below we need this variable in two closures. In rust they prefer explicit cloning

**Pull request merged**